chairman/Chairman.hs

```diff
@@ -0,0 +1,434 @@
+
+-- | Run chairman: connect with all the core nodes.  Chairman will store the
+-- forks from a common prefix.  If any of them is longer than the security
+-- parameter it will throw an exception.
+--
+-- It is also possible to specify how many blocks should be validated.
+--
+runChairman :: forall blk.
+               ( RunNode blk
+               , TraceConstraints blk
+               )
+            => Protocol blk
+            -> [CoreNodeId]
+            -> NumCoreNodes
+            -> Int
+            -- ^ security parameter, if a fork is deeper than it 'runChairman'
+            -- will throw an exception.
+            -> Maybe BlockNo
+            -- ^ finish after that many blocks, if 'Nothing' run continuously.
+            -> Tracer IO String
+            -> IO ()
+runChairman ptcl nids numCoreNodes securityParam maxBlockNo tracer = do
+
+    (chainsVar :: ChainsVar IO blk) <- newTVarM
+      (Map.fromList $ map (\coreNodeId -> (coreNodeId, CF.Empty)) nids)
+
+    void $ flip mapConcurrently nids $ \coreNodeId ->
+        let ProtocolInfo{pInfoConfig} =
+              protocolInfo numCoreNodes
+                           coreNodeId
+                           ptcl
+
+        in connectTo
+            (,)
+            (localInitiatorNetworkApplication
+              coreNodeId
+              chainsVar
+              securityParam
+              maxBlockNo
+              (contramap show tracer)
+              nullTracer
+              nullTracer
+              pInfoConfig)
+            Nothing
+            (localSocketAddrInfo (localSocketFilePath coreNodeId))
+
+
+data ChairmanTrace blk
+  = NotFoundCommonBlock [Maybe blk]
+  | WitnessedConsensusAt SlotNo [Point blk]
+
+instance (Condense blk, Condense (HeaderHash blk)) => Show (ChairmanTrace blk) where
+    show (NotFoundCommonBlock blks)
+      = "not found common block" ++ condense blks
+    show (WitnessedConsensusAt slotNo tips)
+      = mconcat
+      [ "witnessed consensus at slot "
+      , condense slotNo
+      , " current tips: "
+      , condense tips
+      ]
+
+
+--
+-- Shared State, and its API.
+--
+
+
+-- | Shared state between chain-sync clients.  Each chain-sync client will write to the
+-- corresponding entry.
+--
+type ChainsVar m blk = TVar m (Map CoreNodeId (ChainFragment blk))
+
+
+-- | Add a single block to the chain.
+--
+addBlock
+    :: forall blk m.
+       ( MonadSTM m
+       , HasHeader blk
+       )
+    => CoreNodeId
+    -> ChainsVar m blk
+    -> blk
+    -> STM m ()
+addBlock coreNodeId chainsVar blk =
+    modifyTVar' chainsVar (Map.adjust (CF.addBlock blk) coreNodeId)
+
+
+data ChairmanError blk = NodeMisconduct [blk]
+
+instance Condense blk => Show (ChairmanError blk) where
+    show (NodeMisconduct blks) = "NodeMisconduct " ++ condense blks
+
+instance (Condense blk, Typeable blk) => Exception (ChairmanError blk)
+
+
+-- | Check if the oldest immutable tip agree, keep forks from the last common
+-- block (including it).
+--
+checkAndSubtract
```

Comment:
`Subtract` here is a bit confusing (in the helper too). Isn't there a clearer name?

Reply:
what about `prune`?

Reply:
Yeah, that's better 👍